### PR TITLE
Improve output window performance for ant java tasks #4141.

### PR DESCRIPTION
**Pull request by notzed**
These two together improve a test case of 10000 lines of 2890
characters from taking an estimated 25 minutes to run to taking 5
seconds:
 Change to use matches() rather than find() for the stack checking
  regex, which runs very slowly on long lines.
 Use buffered reads rather than single byte reads.

Other changes:
 Use an ExecutorService to start and manage threads.
 Remove references to deprecated ThreadDeath exception.
 Move critical section to isolated functions.
 Split non-processing and processing stream handlers
  into separate classes.
 Fix some logic errors introduced from / and requiring ancient
  'hot fixes'.

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

### CODE REVIEW

**Comment by ebarboni** [comment]
Hi, maybe you should check your name in your git sound like an avatar

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [inline extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:201]: `ifPresentOrElse` is Java 9. Some modules have to stay compatible with Java 8 still, unfortunately. (thats the reason why the build fails)

**Comment by mbien** [comment]
> Hi, maybe you should check your name in your git sound like an avatar

```
git commit --amend --author="Author Name <email@address.com>" --no-edit
```
this would update the author of the last commit in the current branch. After that just force push to the remote branch.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:201]: huh, i guess there's another netbeans bug then because the source level is set to 1.8 and it didn't pick this up.

**Comment by mbien** [comment]
@notzed your github name is fine. I believe Eric meant the name on the commit which is "Not Zed". This has to be a real person I believe, with a correct mail address.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [inline extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:340]: gonna have to check if matches() gives the same results as find() with the given regex.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [inline extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:206]: small nitpick: indentation

**Review by mbien — COMMENTED** [review]
all tests are green. Thats a good sign. Impl looks good at first glance, but I have to take a closer look later - great cleanup for sure.

I believe building NB via shell was known to be faster than via NB. This makes me curious how it performs post patch.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:340]: I'm pretty certain it does because it contains .* on both ends so will always match the whole string anyway.  Also probably more to the point this is just to check if it should ignore the lines because JavaAntLogger will mark them up - and JavaAntLogger uses the same regex and uses matches().

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java]: you can actually configure it on a per-project basis which would override the global default. But NetBeans has ~840 sub projects so its probably not worth the trouble :)

I think the only change I made from the global default is to set tab size to 4 (I believe 8 is default). But this would only show in files which actually use tabs.

I rarely use auto formatting. If i do its usually only for few selected lines since it creates too much noise in the diff.

It was not super important anyway, I only mentioned it since I noticed it while looking through it.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:340]: yes I think you are right. The regex already matches the full line, `find()` would be redundant and likely much slower.
I mostly tried to find an excuse to use the regex checker window #4200.

Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:340]: For the test i was using, find vs matches is roughly ~1500x slower and there are worse cases.  Due to the greedy operator it matches the whole string then walks all the way back to the beginning testing each position looking for the next substring match.  With matches() it then terminates but because it isn't anchored to the start of the string find() advances by one character and repeats - all the way to the other end - so it's particularly bad with long lines, and is the reason for the line length checking here and in javaantlogger. I filed a jdk performance issue against java.util.regex.Pattern but my initial assessment was wrong as to the cause, and it would be tricky to address due to the way the state is managed in the compiled expression.

perl doesn't even execute the regex because it does a pre-scan and determines the string can't possibly match due to the bits that aren't optional which are not present like '('.

Removing the .* at either end makes find() run about the same as matches().  With some simplification and no capturing groups it can run somewhat faster than calling matches() but it's no longer a significant part of the processing time and doesn't seem worth maintaining another equally unreadable expression.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [inline extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: second task should use `errEncoding`.

i think we could also store the encoding as Charset, this would remove lookups inside toString(... "encoding"). 

```diff
diff --git a/extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java b/extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
index b9fbad1..4f91cca 100644
--- a/extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
+++ b/extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
@@ -25,6 +25,7 @@
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.UnsupportedEncodingException;
+import java.nio.charset.Charset;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
@@ -104,8 +105,8 @@
 
     private class NbRedirector extends Redirector {
 
-        private String outEncoding = System.getProperty("file.encoding"); // NOI18N
-        private String errEncoding = System.getProperty("file.encoding"); // NOI18N
+        private Charset outEncoding;
+        private Charset errEncoding;
 
         // #158492. In Ant 1.8.0 output redirection cannot be distinguished by OutputStream subclass type
         // (LogOutputStream vs OutputStreamFunneler$Funnel) as OutputStreamFunneler$Funnel is
@@ -115,6 +116,14 @@
 
         public NbRedirector(Task task) {
             super(task);
+            try {
+                Charset cs = Charset.forName(System.getProperty("file.encoding")); // NOI18N
+                outEncoding = cs;
+                errEncoding = cs;
+            } catch (IllegalArgumentException ex) {
+                outEncoding = Charset.defaultCharset();
+                errEncoding = Charset.defaultCharset();
+            }
         }
 
         public @Override ExecuteStreamHandler createHandler() throws BuildException {
@@ -123,12 +132,12 @@
         }
 
         public @Override synchronized void setOutputEncoding(String outputEncoding) {
-            outEncoding = outputEncoding;
+            outEncoding = Charset.forName(outputEncoding);
             super.setOutputEncoding(outputEncoding);
         }
 
         public @Override synchronized void setErrorEncoding(String errorEncoding) {
-            errEncoding = errorEncoding;
+            errEncoding = Charset.forName(errorEncoding);
             super.setErrorEncoding(errorEncoding);
         }
 
@@ -199,7 +208,7 @@
                     .orElse(null);
                 if (buildLogger != null) {
                     tasks.submit(new MarkupCopier(stdout, Project.MSG_INFO, false, outEncoding, buildLogger, foldingHelper));
-                    tasks.submit(new MarkupCopier(stderr, Project.MSG_WARN, true, outEncoding, buildLogger, foldingHelper));
+                    tasks.submit(new MarkupCopier(stderr, Project.MSG_WARN, true, errEncoding, buildLogger, foldingHelper));
                 } else {
                     setCopier(stdout, getOutputStream(), delegateOutputStream, false);
                     setCopier(stderr, getErrorStream(), delegateErrorStream, true);
@@ -274,7 +283,7 @@
 
         private final InputStream in;
         private final int logLevel;
-        private final String encoding;
+        private final Charset encoding;
         private final RequestProcessor.Task flusher;
         private final ByteArrayOutputStream currentLine;
         private final OutputWriter ow;
@@ -282,7 +291,7 @@
         private final AntSession session;
         private final FoldingHelper foldingHelper;
 
-        public MarkupCopier(InputStream in, int logLevel, boolean err, String encoding, NbBuildLogger buildLogger, FoldingHelper foldingHelper) {
+        public MarkupCopier(InputStream in, int logLevel, boolean err, Charset encoding, NbBuildLogger buildLogger, FoldingHelper foldingHelper) {
             this.in = in;
             this.logLevel = logLevel;
             this.err = err;

```

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [inline extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:368]: i like that you use `foldingHelper` as monitor instead of the task. This might fix a race condition, although interleaving `out` and `err` wouldn't result in proper folding anyway most of the time I suppose.

**Review by mbien — COMMENTED** [review]
found a little issue, nothing major. Rest looks good to me.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: Thanks for spotting the paste-o, i agree on the charset change, i just left it as is to avoid unnecessary changes and since that form is widely used everywhere else in java code and avoids the need to handle the exception.

Do you want me to apply this diff and re-push the pull request?

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:368]: It's a bit ugly but while i was cleaning up the final patch I noticed the comment on FoldingHelper stated it was supposed to be used as a lock for interleaving output, but then it wasn't actually used.  

With this changed it will ensure stderr/stdout has full line outputs before interleaving unless the timeout/flush happens in-between.  The output from the case below is improved although each line still interleaves due to the output rate.

    public class SlowOutput {
    static void deepStackTrace(int i) {
        if (i == 0) {
            try {
                System.out.println("/home/notzed/.emacs:5:4: yo ho");
                throw new Exception();
            } catch (Exception x) {
                x.printStackTrace(System.err);
            }
        } else {
            deepStackTrace(i - 1);
        }
    }
    public static void main(String[] args) throws InterruptedException {
        for (int i = 0; i < 10000; i++) {
            for (int j = 0; j < 1000; j++) {
                System.out.print(j);
            }
            System.out.println();
            if (i % 50 == 49)   deepStackTrace(5);
        }
    }
    }

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: feel free to apply the diff or implement it yourself, whatever you prefer. I simply refactored it to check if it would work and pasted my version.

You can add it to the commit and force push into the PR branch.

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:368]: good stuff!

File: extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java
Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: btw the way i see this: the exception would happen anyway, just later in the toString(... "encoding") method. So, converting it early to `Charset` should be fine without breaking contracts.

Comment (notzed) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: Well now this code has to make a decision on what to do - the only exception being handled (illegalargumentexception) happens only if the property is null, and that case would've previously thrown nullpointerexception in String:lookupCharset().  Either way having a fallback is potentially a behavioural change.

I looked at an alternative approach but then noticed ByteArrayOutputStream:toString(Charset) was introduced at Java 10, so it can't be used anyway can it?

For what it's worth I had looked into a different buffering mechanism previously that could just use String:new directly but it didn't seem worth the added code.

Comment (mbien) [reply extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java:202]: > ByteArrayOutputStream:toString(Charset) was introduced at Java 10, so it can't be used anyway can it?

oh wow you are right. I haven't checked for that. Lets forget about this then - my mistake. Lets only fix the `errEncoding` and we are good.

just as sidenote:
all other exceptions the method throws are subtypes of IllegalArgumentException, so it is basically catching everything, not only the null case (but making this a multi-catch would be clearer probably).

**Review by mbien — APPROVED** [review]

**Comment by mbien** [comment]
planning to merge this soon unless someone speaks up @lkishalmi @matthiasblaesing

**Comment by mbien** [comment]
oh no @notzed ;) don't merge into this branch. Now we have a 
`Author: notzed <51409789+notzed@users.noreply.github.com>` commit in here.

You think you can revert that merge and only keep the commit?
`git reset HEAD^`
should remove the last commit. And simply force push again.

In future if you want to refresh PRs you can do something like this:
`git pull --rebase --autostash origin master`
this is going to put the commit on top of latest master. (assuming origin points at the main repo, but in general its only needed to refresh PRs when they are a few months old or have conflicts)

**Comment by notzed** [comment]
I must've read somewhere to keep the fork up to date so i've been doing that, i wondered why the last time did something funny - the git told me to run some commands so i did and that's what must've done it.

Err I think i fubard something else now.

**Comment by mbien** [comment]
all good. now hands off :)

**Comment by mbien** [comment]
@notzed congrats to your first contribution

btw you can link your commit author details to your account by adding the email address you used for the commit to your github account somewhere in the settings.

**Comment by notzed** [comment]
Goodo, been a while since I contributed much to anything but i was very active in GNOME back in the day (2nd iteration of gnome-terminal, evolution).  Dunno if i'll find something else to poke around in netbeans but we'll see - i quit my job 18 months ago and can't be bothered getting another one yet so i've got a lot of time on my hands and it's getting a bit boring.

**Pull request merged**
